plugins/utils/which-key.nix

```diff
@@ -55,257 +265,308 @@ with lib;
+    preset = false;
 
-      warnings = optional opt.registrations.isDefined ''
-        nixvim (plugins.which-key):
-        The option definition `plugins.which-key.registrations' in ${showFiles opt.registrations.files} has been deprecated in which-key v3; please remove it.
-        You should use `plugins.which-key.settings.spec' instead.
+    delay = 200;
 
-        Note: the spec format has changed in which-key v3
-        See: https://github.com/folke/which-key.nvim?tab=readme-ov-file#%EF%B8%8F-mappings
-      '';
+    spec = [
+      {
+        __unkeyed = "<leader>b";
+        group = "󰓩 Buffers";
+      }
+      {
+        __unkeyed = "<leader>bs";
+        group = "󰒺 Sort";
+      }
+      {
+        __unkeyed = "<leader>g";
+        group = "󰊢 Git";
+      }
+      {
+        __unkeyed = "<leader>f";
+        group = " Find";
+      }
+      {
+        __unkeyed = "<leader>r";
+        group = " Refactor";
+      }
+      {
+        __unkeyed = "<leader>t";
+        group = " Terminal";
+      }
+      {
+        __unkeyed = "<leader>u";
+        group = " UI/UX";
+      }
+    ];
 
-      extraConfigLua =
-        ''
-          require("which-key").setup(${helpers.toLuaObject setupOptions})
-        ''
-        + (optionalString (opt.registrations.isDefined && cfg.registrations != { }) ''
-          require("which-key").register(${helpers.toLuaObject cfg.registrations})
-        '');
+    replace = {
+      desc = [
+        [
+          "<space>"
+          "SPACE"
+        ]
+        [
+          "<leader>"
+          "SPACE"
+        ]
+        [
+          "<[cC][rR]>"
+          "RETURN"
+        ]
+        [
+          "<[tT][aA][bB]>"
+          "TAB"
+        ]
+        [
+          "<[bB][sS]>"
+          "BACKSPACE"
+        ]
+      ];
     };
+
+    notify = false;
+
+    win = {
+      border = "single";
+    };
+
+    expand = 1;
+  };
+
+  # TODO: introduced 2024-08-05: remove after 24.11
+  extraOptions = {
+    registrations = mkOption {
+      type = with types; attrsOf anything;
+      description = ''
+        This option is deprecated, use `settings.spec` instead.
+
+        Note: the keymap format has changed in v3 of which-key.
+      '';
+      visible = false;
+    };
+  };
+
+  extraConfig = cfg: {
+
+    # TODO: introduced 2024-08-05: remove after 24.11
+    warnings = optional opt.registrations.isDefined ''
+      nixvim (plugins.which-key):
+      The option definition `plugins.which-key.registrations' in ${showFiles opt.registrations.files} has been deprecated in which-key v3; please remove it.
+      You should use `plugins.which-key.settings.spec' instead.
+
+      Note: the spec format has changed in which-key v3
+      See: https://github.com/folke/which-key.nvim?tab=readme-ov-file#%EF%B8%8F-mappings
+    '';
+
+    extraConfigLua = optionalString (opt.registrations.isDefined && cfg.registrations != { }) ''
+      require("which-key").register(${toLuaObject cfg.registrations})
+    '';
+  };
```

Comment:
We can wrap the _whole thing_ in optionalAttrs now, since the entire config block relates to the deprecated `registrations` option:

```nix
  # TODO: introduced 2024-07-29: remove after 24.11
  # NOTE: this may be upgraded to a mkRemoveOptionModule when which-key removes support
  extraConfig =
    cfg:
    let
      # I'm on the fence whether this is better off here or at the top of the file
      # If it's here, then we can remove this whole block in one go when removing the deprecated option, I guess...
      opt = options.plugins.which-key;
    in
      optionalAttrs opt.registrations.isDefined {
        warnings = [
          ''
            nixvim (plugins.which-key):
            The option definition `plugins.which-key.registrations' in ${showFiles opt.registrations.files} has been deprecated in which-key v3; please remove it.
            You should use `plugins.which-key.settings.spec' instead.

            Note: the spec format has changed in which-key v3
            See: https://github.com/folke/which-key.nvim?tab=readme-ov-file#%EF%B8%8F-mappings
          ''
        ];

        extraConfigLua = mkIf (cfg.registrations != { }) ''
          require("which-key").register(${toLuaObject cfg.registrations})
        '';
      };
```

Reply:
Doing this is what causes the infinite recursion.

Reply:
It should be fine in this instance, since we aren't defining anything other than `warnings` and `extraConfigLua`

Reply:
I tried and it threw the recursion error. I assumed it had to do with referencing `cfg.` within the optionalAttrs

Reply:
It seems relying on `options` to define the _shape_ of the config definition is infinitely recursive. I.e. `{}` vs `{ warnings = []; }` is a different shape, because warnings isn't defined at all in the first example.

This is a constant shape, I've tried it out locally and it works:

```nix
  # TODO: introduced 2024-07-29: remove after 24.11
  # NOTE: this may be upgraded to a mkRemoveOptionModule when which-key removes support
  extraConfig = cfg: {
    warnings = lib.optional opt.registrations.isDefined ''
      nixvim (plugins.which-key):
      The option definition `plugins.which-key.registrations' in ${showFiles opt.registrations.files} has been deprecated in which-key v3; please remove it.
      You should use `plugins.which-key.settings.spec' instead.

      Note: the spec format has changed in which-key v3
      See: https://github.com/folke/which-key.nvim?tab=readme-ov-file#%EF%B8%8F-mappings
    '';

    extraConfigLua =
      if opt.registrations.isDefined then
        lib.mkIf (cfg.registrations != { }) ''
          require("which-key").register(${toLuaObject cfg.registrations})
        ''
      else
        lib.mkIf false "";
  };
```

This would also work, but feels even messier:

```nix
  extraConfig =
    cfg:
    if opt.registrations.isDefined then
      {
        warnings = [
          # omitted
        ];
        extraConfigLua = lib.mkIf (cfg.registrations != { }) ''
          # omitted
        '';
      }
    else
      {
        warnings = [ ];
        extraConfigLua = lib.mkIf false "";
      };